test/integration/compose/application-manager.spec.ts

```diff
@@ -327,6 +329,370 @@ describe('compose/application-manager', () => {
+					labels,
+					appId: 1,
+					serviceName: 'new',
+					commit: 'new-release',
+					releaseId: 2,
+				}),
+			];
+			return { currentServices, targetServices };
+		};
+
+		it('should not infer a kill step for current service A before target service B download finishes with download-then-kill', async () => {
+			const labels = {
+				'io.balena.update.strategy': 'download-then-kill',
+			};
+			const { currentServices, targetServices } =
+				await createCurrentAndTargetServicesWithLabels(labels);
+
+			/**
+			 * Before target image finishes downloading, do not infer kill step
+			 */
+			let targetApps = createApps(
+				{
+					services: targetServices,
+				},
+				true,
+			);
+			const c1 = createCurrentState({
+				services: currentServices,
+				networks: [DEFAULT_NETWORK],
+				downloading: ['image-new'],
+			});
+			const steps = await applicationManager.inferNextSteps(
+				c1.currentApps,
+				targetApps,
+				{
+					downloading: c1.downloading,
+					availableImages: c1.availableImages,
+					containerIdsByAppId: c1.containerIdsByAppId,
+				},
+			);
+			// There should be two noop steps, one for target service which is still downloading,
+			// and one for current service which is waiting on target download to complete.
+			expectSteps('noop', steps, 2);
+			// No kill step yet
+			expectNoStep('kill', steps);
+
+			/**
+			 * After target image finishes downloading, infer a kill step
+			 */
+			targetApps = createApps(
+				{
+					services: targetServices,
+				},
+				true,
+			);
+			const { currentApps, availableImages, downloading, containerIdsByAppId } =
+				createCurrentState({
+					services: currentServices,
+					networks: [DEFAULT_NETWORK],
+					images: [
+						// Both images have been downloaded or had their metadata updated
+						createImage({
+							name: 'image-main',
+							appId: 1,
+							serviceName: 'main',
+							commit: 'new-release',
+							releaseId: 2,
+						}),
+						createImage({
+							name: 'image-new',
+							appId: 1,
+							serviceName: 'new',
+							commit: 'new-release',
+							releaseId: 2,
+						}),
+					],
+				});
+			const steps2 = await applicationManager.inferNextSteps(
+				currentApps,
+				targetApps,
+				{
+					downloading,
+					availableImages,
+					containerIdsByAppId,
+				},
+			);
+			// Service `old` is safe to kill after download for `new` has completed
+			const [killStep] = expectSteps('kill', steps2, 1);
+			expect(killStep)
+				.to.have.property('current')
+				.that.deep.includes({ serviceName: 'old' });
+			// Service `new` in target release should be started as download has completed
+			const [startStep] = expectSteps('start', steps2, 1);
+			expect(startStep)
+				.to.have.property('target')
+				.that.deep.includes({ serviceName: 'new' });
+			// No noop steps
+			expectNoStep('noop', steps2);
+		});
+
+		it('should not infer a fetch step for target service B before current service A kill with kill-then-download', async () => {
```

Comment:
why not? A and B are different services, I think it's OK to start fetching B while A is being killed. If anything I would expect that A should not be killed at all before B has been fetched

Reply:
The test case here and the one in the comment below are for `kill-then-download` and `delete-then-download` strategies. Does this still apply? The intent here is for `kill|delete-then-download` to apply app-wide, meaning that all service should be killed/deleted before download occurs, even for services only in target and not in current.

Reply:
Never mind, I see your last comment below which addresses this, we can continue the conversation there